### PR TITLE
EOS-28430 btree: btree-ut hangs when code compiled without --enable-debug flag

### PR DESCRIPTION
**Pull request by kanchan-chaudhari**
Solution:
Make variable "thread_run" volatile.
Make variable struct m0_atomic64::long a_value volatile.

Signed-off-by: Kanchan Chaudhari <kanchan.chaudhari@seagate.com>

# Problem Statement
- "btree-ut" hang when code is compiled without "--enable-debug" flag

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Review by shashank-parulekar — APPROVED** [review]

**Review by ydb242 — APPROVED** [review]

**Review by RadhaGulhane13 — APPROVED** [review]

File: btree/btree.c [hunk 1]
```diff
@@ -9813,8 +9813,9 @@ static struct m0_atomic64 threads_running;
 		bool try_again;                                               \
 		do {                                                          \
 			try_again = false;                                    \
-			if (m0_atomic64_cas(&thread_run, true, false)) {      \
-				while (m0_atomic64_get(&threads_quiesced) <    \
+			if (m0_atomic64_cas((int64_t *)&thread_run, true,     \
```
Comment (upendrapatwardhan) [inline btree/btree.c:9816]: Please assign/check `0,1` instead of `true/false` for int64_t variable.

Comment (kanchan-chaudhari) [reply btree/btree.c]: Done.

**Review by shashank-parulekar — APPROVED** [review]

**Review by upendrapatwardhan — APPROVED** [review]

File: lib/user_space/user_x86_64_atomic.h [hunk 2]
```diff
@@ -38,7 +38,7 @@
  */
 
 struct m0_atomic64 {
-	long a_value;
+	volatile long a_value;
```
Comment (nikitadanilov) [inline lib/user_space/user_x86_64_atomic.h:41]: Why is this needed, please explain.

Comment (shashank-parulekar) [reply lib/user_space/user_x86_64_atomic.h:41]: If we do not do this and call m0_atomic64_get() in a loop then gcc with -O2 flag optimizes this statement to read from the register variable instead of reading from the memory location (which it does correctly with -O0 flag).

Comment (ivan-alekhin) [reply lib/user_space/user_x86_64_atomic.h:41]: @shashank-parulekar , have you tried adding a memory barrier after the m0_atomic64_get() call to solve this issue? It should be enough to give the compiler a hint, so that it would re-load the variable instead of caching it in the registers.
